pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -46,35 +128,36 @@ func Endpoints(ds *v1.DiscoveryService, serviceClusters []string) *xdsapi.Discov
 		// responses. Pilot believes in eventual consistency and that at some point, Envoy
 		// will begin seeing results it deems to be good.
 		VersionInfo: version,
-		Nonce:       version,
+		Nonce:       version + "-" + time.Now().String(),
 	}
 
-	var totalEndpoints uint32
 	out.Resources = make([]types.Any, 0, len(serviceClusters))
 	for _, serviceCluster := range serviceClusters {
-		hostname, ports, labels := model.ParseServiceKey(serviceCluster)
-		instances, err := ds.Instances(hostname, ports.GetNames(), labels)
-		if err != nil {
-			log.Warnf("endpoints for service cluster %q returned error %q", serviceCluster, err)
-			continue
-		}
-		locEps := localityLbEndpointsFromInstances(instances)
-		if len(instances) == 0 {
-			log.Infoa("EDS: no instances ", serviceCluster, hostname, ports, labels)
+		clAssignmentRes := s.clusterEndpoints(ds, serviceCluster)
+		if clAssignmentRes != nil {
+			out.Resources = append(out.Resources, *clAssignmentRes)
 		}
-		clAssignment := &xdsapi.ClusterLoadAssignment{
-			ClusterName: serviceCluster,
-			Endpoints:   locEps,
-		}
-		//log.Infof("EDS: %v %s", serviceCluster, clAssignment.String())
-		clAssignmentRes, _ = types.MarshalAny(clAssignment)
-		out.Resources = append(out.Resources, *clAssignmentRes)
-		totalEndpoints += uint32(len(locEps))
 	}
 
 	return out
 }
 
+// Get the ClusterLoadAssignment for a cluster.
+func (s *DiscoveryServer) clusterEndpoints(ds *v1.DiscoveryService, serviceCluster string) *types.Any {
+	c := s.getOrAddEdsCluster(serviceCluster)
+	if c.LoadAssignments == nil { // fresh cluster
+		updateCluster(serviceCluster, c)
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```

Comment:
RLock and RUnlock

Reply:
Doesn't seem to exist in 1.9 ?

Reply:
As Shriram said, RWMutex. Also, we've moved over to Go 1.10 already, haven't we? We certainly compile istio/api with golang 1.10.